### DISCUSSION TITLE
用 lua 实现小鹤音形官方版「仅在空码时显示后续编码」功能

### DISCUSSION THREAD

**Discussion by amorphobia** [amorphobia/openfly#33] [展示/Show and tell]
官方版小鹤音形并没有一个词库用于内嵌的编码提示，但是开源小鹤有一个[词典](https://github.com/amorphobia/openfly/blob/3d667e21c63d592b482e54b52a54bcbb7041707c/openfly.embedded.hint.dict.yaml)，配合 [lua 脚本](https://github.com/amorphobia/openfly/blob/3d667e21c63d592b482e54b52a54bcbb7041707c/lua/openfly_hint_filter.lua)用于这个提示。

这里展示一个方案，使用了 lua 脚本来实现自动的提示，但是需要对词典进行部分甚至是大量的改动。

### 配置的改动
这个方法大致的思路就是使用 Rime 自身的自动补全，判定在空码的时候，将自动补全的候选排序第一的词语展示出来，因此需要将 `translator/enable_completion` 改为 `true`。

另外如果要实现排序，则需要在 `engine/filters` 里面加上 `lua_filter@openfly_empty_code_completion_filter`. 当然这个过滤器名字可以自定义。

如果只是实现空码时自动提示一个候选，那么这样改动就可以了，若想自动提示 `n` 个候选，那么需要添加一个选项 `empty_code_completion_size: n`，其中 `n` 是一个整数。

### 词典的改动
对于词典，需要给部分词加上排序信息，数字越小越优先。如果不改动词典，也可以自动补全，但候选的优先级就是 Rime 引擎决定的，[无法改动](https://github.com/rime/librime/issues/297#issuecomment-511084325)。

改动的例子如下：
```
收买人心`[1]	umrx
```
在词的后面加上 `` `[排序]``，这里的排序数字越小就越靠前。小鹤音形官方版默认会在空码时显示一个自动提示，如果仅为了实现这个功能，最多只需要在 [primary 词典](https://github.com/amorphobia/openfly/blob/3d667e21c63d592b482e54b52a54bcbb7041707c/openfly.primary.dict.yaml)里给 [embedded.hint 词典](https://github.com/amorphobia/openfly/blob/3d667e21c63d592b482e54b52a54bcbb7041707c/openfly.embedded.hint.dict.yaml)有的词的词加上排序信息，注意是在 primary 词典里加。其实也不需要给所有在 embedded.hint 词典里的词语都加上，因为有的空码本身就只会有一个候选，抑或是有多个候选，但第一优先正是需要的词。

如果要实现自动补全任意个数并且排序，则需要给相应的词语都加上排序信息。去看小鹤音形官方版的码表可以发现，甚至所有的词语都有一个「排序」项，而且不能导出，这也是本方法不会发布到仓库里的直接原因——工作量太大了，无法维护。

### Lua 脚本
主要脚本 `openfly_empty_code_completion_filter.lua` 放在 `lua` 文件夹里：
```lua
local common = require("openfly_common")

local function filter(input, env)
  local skip = env.engine.context.input:sub(1,1) == "`"
  local candidates = {}
  local index_text = {}
  local empty_code = true
  for cand in input:iter() do
    if not skip then
      table.insert(index_text, cand.text)
      candidates[cand.text] = cand
      if cand.type ~= "completion" then
        empty_code = false
      end
    else yield(cand) end -- 反查时保持候选不变
  end
  if not empty_code then -- 非空码时忽略自动补全
    for i, t in pairs(index_text) do
      local cand = candidates[t]
      if cand ~= nil then
        local comment = string.gsub(cand:get_genuine().comment, "%[%d+%]", "")
        if cand.type ~= "completion" then
          yield(Candidate(cand.type, cand.start, cand._end, cand.text, comment))
        elseif comment ~= nil and comment:sub(-1):match("[^%w/%]]") ~= nil then
          -- 假设补全选项的注释都以字母或 "/" 结尾，以此排除 xmoji 的条目
          yield(cand)
        end
      end
    end
  else -- 空码情况
    local sorted_cand = {}
    for i, t in pairs(index_text) do
      local cand = candidates[t]
      if cand ~= nil then
        local note = tostring(cand:get_genuine().comment:match("%[%d+%]"))
        local id = tonumber(note:sub(2, -2))
        if id ~= nil then
          table.insert(sorted_cand, { idx = id, cnd = cand })
        else
          table.insert(sorted_cand, { idx = 100000, cnd = cand })
        end
      end
    end
    table.sort(sorted_cand, function(a, b) return a.idx < b.idx end)
    for i in pairs(sorted_cand) do
      local lmt = common.rime_option.get_int(env, "empty_code_completion_size") or 1
      if i > lmt then break end
      local cand = sorted_cand[i].cnd
      local comment = string.gsub(cand:get_genuine().comment, "%[%d+%]", "")
      yield(Candidate(cand.type, cand.start, cand._end, cand.text, comment))
    end
  end
end

return filter
```

在 `rime.lua` 文件里加上一行 `openfly_empty_code_completion_filter = require("openfly_empty_code_completion_filter")`, 这样就行了。

**Comment by selong1234** [comment]
大佬 我用了你的方法实现了官方版的功能
有两个小问题想请教一下
1. 打o的时候会出来很多 符号那些 而不只是出“哦”
2. 打空码时会出现提示，例如打aab，会出现阿爸~a，我在代码里面把comment = “”，暂时关闭了提示，我不太懂lua，不知道这样改会不会影响别的地方，想问一下开源小鹤有没有能控制开关提示的地方？
